### DISCUSSION TITLE
What's the story about the included PnP Library?

### DISCUSSION THREAD

**Discussion by TullettJay** [Zerg00s/FlowPowerAppsMigrator#6] [Q&A]
Hi:

I was a bit surprised to see that you had included a copy of the PnP PowerShell library with the application, complete with DLLs.

What led you to use this copy instead of relying on the latest release of PnP PowerShell?

Best Regards,
lonesome-cloner

**Comment by Zerg00s** [comment]
Hi @lonesome-cloner, 

### The reason I wanted to include the module with the tool is to
- Make it portable, self-contained.
- Reduce number of steps it takes for you to start using this tool.

I see many people who I work with having issues even with the simple act of installing a PowerShell module. Even very technical colleagues struggle sometimes. To avoid dealing with these issues, I've just included all dependencies inside the whole package.

### Why I included the older PnP rather than the new one
I wanted this tool to be portable ("sendable" over an email). For this reason, I actually wanted to included an even older version of PnP, from around 2018 (might still do it later). The older version used to be very lightweight, 30 MB, then it grew to 55 MB (currently used by this project). Then the modern PnP.PowerShell is around 80MB. Maybe I can simply delete some of the DLL's that I know for a fact are not used by the module? I need to give it a try.

The fact that this tool imports a PowerShell module does not interfere in any way with other modules you might already have installed locally.

**Reply by mchestnutthreewill** [reply]
I, too, was wondering why Import-Module was used to import the included SharePointPnPPowerShellOnline v3.29.2101.0.

This comment in RunAllScripts.ps1 caught my attention: 
`# Legacy PowerShell PnP Module is used because the new one has a critical bug`

But, it wasn't clear in anything I could find what the critical bug is or which version of the "new one" had the bug.

I just ran a test using PnP.PowerShell v1.9 and did not see any issues with the generation of the resulting dist/.zip. 
- I extracted the contents of the src/dist .ZIP and unzipped the Forms .msapp, too
- I ran [prettier](https://prettier.io/docs/en/install.html) to format the extracted JSON to make the comparison easier
- I compared the folders using [WinMerge](https://winmerge.org), all files were identical
  - The only difference was the binary .msapp file. 
  - The extracted contents of the .msapp were identical, but the WinMerge said the binary .msapp file differed from the original
  - But, I got the same "diff" if I used SharePointPnPPowerShellOnline v3.29.2101.0 twice in row and compared the results from each. All extracted JSON files were the same (even those extracted from the .msapp) but the .msapp itself showed differences.
- This tells me that the same output is generated if either SharePointPnPPowerShellOnline v3.29.2101.0 or PnP.PowerShell v1.9 is used

**Comment by Zerg00s** [comment]
Hi @mchestnutthreewill Thank you very much for testing it. I am planning to upgrade to the latest PnP.PowerShell module.  I tried it a times over the last year, it just never worked for me. 

I didn't try the PnP.PowerShell v1.9 yet, but there is a good chance it has the same issue as the previous version since the change log does not mention a fix.

I don't have a specific list of steps for reproduction, but from what I remember, I could never export SharePoint lists and Import them using PnP.PowerShell. Below is a quick example from memory. The  Export/Import process just never worked for me. I saw several issues reported but saw no solution for this. Unfortunately, I was also too lazy to report it myself, which I regret now.

## Test of PnP.PowerShell 
```
# Connect to Site A with one of the two methods
Connect-PnPOnline -Url "https://contoso.sharepoint.com" -UseWebLogin
# Connect-PnPOnline -Url "https://contoso.sharepoint.com" -Interactive

Get-PnPSiteTemplate -Handlers Lists -Out MyLists.xml -ListsToExtract "Title of List One","Title of List Two"

# Connect to site B with one of the two methods
Connect-PnPOnline -Url "https://tailspin.sharepoint.com" -UseWebLogin
# Connect-PnPOnline -Url "https://tailspin.sharepoint.com" -Interactive
Apply-PnPProvisioningTemplate .\MyLists.xml
```

**Reply by bhlaws** [reply]
I, for one, am glad this project is still using the legacy PnP module. My organization is unable to upgrade to the cross-platform module because of how they changed the unattended sign-in process. It's no longer possible to provide a simple login/password and instead requires an AAD App Registration. That's hard to get in my org, so we haven't done it yet. Anyway, I selfishly support using the old module in this project. :-D
